torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+        for global_group, local_group in zip(self.param_groups, groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        # Check whether we got a local or global dict
+        if state_dict["local_state_dict"]:
+            self.load_local_state_dict(state_dict)
+        else:
+            # Get this optimizer's param_groups shard
+            param_groups = state_dict["param_groups"][
+                state_dict["partition"][self.rank][0] : state_dict["partition"][self.rank][1]
+            ]
+            # Dispatch this rank's state dictionary to the wrapped shard optimizer
+            self.load_local_state_dict({"state": state_dict["state"][self.rank], "param_groups": param_groups})
+
+    def add_param_group(self, param_group: Dict) -> None:
+        """Add a param group to the :class:`Optimizer` s ``param_groups``.
+
+        This can be useful when fine tuning a pre-trained network as frozen layers can be made
+        trainable and added to the :class:`Optimizer` as training progresses.
+
+        Arguments:
+            param_group (dict): Specifies what Tensors should be optimized along with group
+            specific optimization options
+
+        .. warning: This handles updating the shards on all partitions, but needs to be called on all ranks.
+        """
+
+        super().add_param_group(param_group)
+        if not self.in_super_constructor:
+            # Force a re-partitioning
+            self._partition_parameters.clear()
+            self._per_device_params.clear()
+            self._param_rank.clear()
+
+            param_groups = self.partition_parameters()[self.rank]
+            if len(param_groups) == len(self.optim.param_groups) + 1:
+                self.optim.add_param_group(param_groups[-1])
+
+    ##########################
+    # Private helper functions
+    ##########################
+
+    def _broadcast_params(self) -> None:
+        """Helper function to broadcast all the parameters from a given device"""
+
+        with torch.no_grad():
+            # All the params on this device (inc all ranks)
+            for (
+                device,
+                device_params,
+            ), bucket_list in zip(self.per_device_params.items(), self._buckets.values()):
+
+                # Go through all the ranks and broadcast the new parameters, bucket the smallest ones
+                for (src_rank, params), bucket in zip(enumerate(device_params), bucket_list):
+
+                    global_src_rank = self.get_global_rank(self.group, src_rank)
```

Comment:
Because it's being used in conjunction with model parallel or pipelining for instance, so there can be sub-groups and not all ranks will cooperate. Fairseq is actively doing that